### PR TITLE
Full MSVC support and bug fixes.

### LINKED ISSUE #18: Support msvc
### Search before asking

- [x] I searched the [issues](https://github.com/alibaba/yalantinglibs/issues) and found no similar issues.

### What happened + What you expected to happen
easylog compile failed with msvc, should be fixed.

### Reproduction way
build with msvc2022

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by refvalue**
## Why

Closes #18 
Made this repo available to MSVC.

## What is changing

Added:
- util/magic_names.hpp
- util/meta_string.hpp
- A uniform MSVC/GCC/Clang implementation of reflecting qualified function names.
- A meta_string class for template programming as non-type parameter when enabling C++20.

Added:
- util/benchmark
- util/examples
- util/tests
Some static asserts for ensuring correctness of newly added code.

Modified/Fixed:
- coro_rpc: CMakeLists.txt
- coro_rpc/remote.hpp
- util/function_name.h
Fixed fatal issues when compiling with MSVC.
Replaced -pthread with the Threads::Threads CMake package to eliminate MSVC warnings.
Fixed incorrect function names when using MSVC.

All unit tests have passed on MSVC 19.3x/GCC11.x/12.x/Clang 15+.
Core logics have passed on GoldBolt.

## Example
Please view src/util/tests

### CODE REVIEW

**Comment by poor-circle** [comment]
Thank you very much!

**Comment by PikachuHyA** [comment]
good job

File: src/util/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+message("The \"util\" project does not need a main project.")
```
Comment (PikachuHyA) [inline src/util/CMakeLists.txt:1]: please use `message(STATUS message_string)`
```
message(STATUS "The \"util\" project does not need a main project.")
```

File: src/util/benchmark/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1 @@
+message("The \"util\" project does not need a benchmark.")
```
Comment (PikachuHyA) [inline src/util/benchmark/CMakeLists.txt:1]: please use `message(STATUS message_string)`

File: src/util/examples/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1 @@
+message("The \"util\" project does not contain examples. Please read the \"tests\" for your information.")
```
Comment (PikachuHyA) [inline src/util/examples/CMakeLists.txt:1]: please use `message(STATUS message_string)`

File: src/coro_rpc/benchmark/client/CMakeLists.txt [hunk 4]
```diff
@@ -3,7 +3,7 @@ target_link_libraries(benchmark_client PRIVATE libcoro_rpc)
 target_include_directories(benchmark_client PRIVATE . ..)
 
 add_executable(benchmark_data_gen data_gen.cpp)
-target_link_libraries(benchmark_data_gen PRIVATE libcoro_rpc)
+target_link_libraries(benchmark_data_gen PRIVATE libcoro_rpc Threads::Threads)
```
Comment (PikachuHyA) [inline src/coro_rpc/benchmark/client/CMakeLists.txt:6]: I have fixed the `-pthread` problem by https://github.com/alibaba/yalantinglibs/pull/30 yesterday.
please rebase to the latest main branch.

**Comment by qicosmos** [comment]
Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/alibaba/yalantinglibs?pullRequest=35) before we can accept your contribution.
@refvalue

**Comment by PikachuHyA** [comment]
The CI is added by https://github.com/alibaba/yalantinglibs/pull/33. However, only `struct_pack` enabled on MSVC.
Please change the `yaLanTingLibs_ENABLE_PROJECTS=all` to enable all projects on MSVC and verify your code.
https://github.com/alibaba/yalantinglibs/blob/657718f346c66e1c4dcb1165531b2e4c6c3b75af/.github/workflows/windows.yml#L42
note: Windows Server 2022 is used in CI.

File: src/util/tests/static_asserts.cpp [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+                meta_string{"func_template<int,double>"});
+#else
+  static_assert(name_of_v<&func_template<int, double>> ==
+                meta_string{"func_template<int, double>"});
+#endif
+}
+}  // namespace ns1::ns2::ns3
+}  // namespace refvalue::tests
+
+int main() {}
```
Comment (PikachuHyA) [inline src/util/tests/static_asserts.cpp:144]: add a newline is better 😊

**Review by PikachuHyA — CHANGES_REQUESTED** [review]
Thanks for your work. 
Please rebase to the latest main branch, and pass the CI.

**Comment by refvalue** [comment]
> Thanks for your work. Please rebase to the latest main branch, and pass the CI.

OK. I will create a new PR later.

**Comment by refvalue** [comment]
> Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/alibaba/yalantinglibs?pullRequest=35) before we can accept your contribution. @refvalue

Thanks! I has forked a new repo with a new PR, for the previous fork had an incorrect email. Now everything goes well.

**Comment by refvalue** [comment]
> Thank you very much!

Thank you!

**Pull request closed without merging**